### ISSUE TITLE
[TRACKER] FBX Importer bugs that depend on Assimp.

### ISSUE THREAD

**Issue by reduz** [godotengine/godot#30800]
This is the list of issues on Assimp's side that we are pending on in order to have proper FBX import in Godot. We are tracking them in one of our issues in order to be on the same page.

- [x] https://github.com/assimp/assimp/issues/2474 - Assimp removes bones not linked to vertices. Godot needs those bones. There was an option added to Assimp to not remove them but it seems broken.
- [ ] https://github.com/assimp/assimp/issues/2498 Assimp preserves FBX pivots and hands them over to us. We don't want them and believe Assimp should collapse them into regular transforms. There was an option added to do this on their side, but it seems broken.
- [x] https://github.com/assimp/assimp/issues/2561 CM to M conversion seems incomplete in Assimp: Not everything is properly converted, resulting in broken files. It also seems that axis conversion does not apply to  those either (needs to be verified).

Will add more as we find them. Feel free to test FBX import in Godot and help us open issues in Assimp to aid their developers in improving the format support.

**Comment by RevoluPowered** [comment]
I've made a godot project for testing this out.

Here is some test models which can only be used for this purpose of FBX importing.

https://github.com/RevoluPowered/GodotEngine.TestModels

Unfortunately I have no more time to test (today) but the project provides 3 animated meshes and UV textures.

It also includes two static meshes without animations. .blend files are present too.

Happy hacking!

**Comment by RevoluPowered** [comment]
> Assimp removes bones not linked to vertices. Godot needs those bones. There was an option added to Assimp to not remove them but it seems broken.

Still a problem with most recent master branch 3.2.

Another issue, bone scales do not come through properly and it seems to be independently scaling the bones, and ignoring the scale, rotation when importing the fbx.

Another in the same thing the skeleton of a mesh becomes detatched even when bones are joined.

**Model ana.blend (blender 2.8 fbx export)**
![image](https://user-images.githubusercontent.com/748770/61864927-407be900-aeca-11e9-946c-1088e5ab7b97.png)

**Comment by RevoluPowered** [comment]
Additionally when playing an animation because of the broken bone scale they crush the model.

![image](https://user-images.githubusercontent.com/748770/61865065-78832c00-aeca-11e9-8dfe-1e32b1733408.png)

**Comment by fire** [comment]
@RevoluPowered Currently opening my latest work on the FBX import. It has diverged from master. I'll give some screenshots.

![editor_screenshot_2019-07-25T083856-0700](https://user-images.githubusercontent.com/32321/61888056-af9c1200-aeb7-11e9-856d-79b192c82f78.png)

The scale is 0.01 different for the mesh and the skeleton. That is why https://github.com/assimp/assimp/issues/2561 also needs to be completed for cm to m.

There's a difference in how the hole filling strategy works between master and my current.

https://github.com/fire/godot/tree/fbx_import_opt

**Comment by RevoluPowered** [comment]
#30822 fbx importer related bugs

**Comment by itnufc** [comment]
Export from 3ds max scale differs. Export im meters, imports in cm.

**Comment by itnufc** [comment]
Export from 3ds max axis swap, Changing y scale value in the godot inspector scales along the z axis and changing z scale value in the godot inspector scales along the y axis .

**Comment by RevoluPowered** [comment]
1 and 3 is fixed.

**Comment by RevoluPowered** [comment]
Working CM to M example NVIDIA demo

 
![image](https://user-images.githubusercontent.com/748770/63723838-59234a00-c84e-11e9-8502-7bed24b5f308.png)

**Comment by RevoluPowered** [comment]
assimp/assimp#2498 we are working on fixing this right now.

Maya uses this explicitly so to get the pipeline from Maya FBX -> Godot to work, it is a blocker.

**Comment by RevoluPowered** [comment]
https://github.com/RevoluPowered/assimp/commit/5193532627e1010462265708a7939e1dc0f97fa7

**Comment by IuryBorges1993** [comment]
Facial animation on Godot results in strange deformation.

in blender it works all normal.

in godot the eyes do not close and the mouth is giving a strange deformation.
![Captura de tela de 2020-01-10 10-12-48](https://user-images.githubusercontent.com/39449626/72155922-fb257800-3392-11ea-8f85-b53563d89320.png)

**Comment by fire** [comment]
As far as I know pivots are correctly implemented in https://github.com/godotengine/godot/pull/39418, but locators aren't. So that completes the remaining issue in this tracker.
